drivers/led/lp50xx.c

```diff
@@ -172,12 +182,15 @@ static int lp50xx_set_color(const struct device *dev, uint32_t led,
 		return -EINVAL;
 	}
 
-	buf[0] = LP50XX_OUT0_COLOR(config->num_modules);
-	buf[0] += LP50XX_COLORS_PER_LED * led_info->index;
-
-	buf[1] = color[0];
-	buf[2] = color[1];
-	buf[3] = color[2];
+	uint8_t const buf[4] = {
+		(
+			LP50XX_OUT0_COLOR(config->num_modules) +
+			LP50XX_COLORS_PER_LED * led_info->index
+		),
+		color[led_info->color_mapping[0] - 1],
+		color[led_info->color_mapping[1] - 1],
+		color[led_info->color_mapping[2] - 1],
```

Comment:
I think there is a misunderstanding here about how color mapping works. The previous code was correct. In `color_mapping`, the array index gives the channel number and array value a color ID that can be used by the user to populate the `color` array. For any color ID greater than 3 (yellow color for example) the code you wrote overflows the color register of the next LED.

Reply:
> For any color ID greater than 3 (yellow for example) the code you wrote overflows the color register of the next LED.

Well, technically, the "runtime validation" in the init function addresses this concern.  I'd love for it to be compile time, but it's C 🤣.  This creates it's own problems though, namely preventing the user from using anything OTHER than RGB.

Reply:
Well, actually the verification code was too restrictive as well. There is nothing which prevents to connect non-RGB (i.e. different colors), or even single or dual-color LEDs to a lp50xx compatible controller.

And the current implementation of the `lp50xx_set_color` function is not correct:

```
static int lp50xx_set_color(const struct device *dev, uint32_t led,
                            uint8_t num_colors, const uint8_t *color)
{
...
        buf[1] = color[0];
        buf[2] = color[1];
        buf[3] = color[2];
```

It is not correct to assume there are 3 elements stored in the `color` array. It could be single-color LEDs...